### DISCUSSION TITLE
Best way to update an existing indicator when you receive a new candle

### DISCUSSION THREAD

**Discussion by SimoneLocatelli** [DaveSkender/Stock.Indicators#883] [Ideas]
Let's suppose I create the following indicator

```
 List<IQuote> candles = GetAllCandles();
 emaFast = candles.GetEma(21); 
```

Currently, when I add a new candle to list, I can update the indicator by recalculating it again
```
candles.Candle(newCandle);
emaFast = candles.GetEma(21); 
```

I was wondering if this is best approach or if there is a more faster/better way to update the indicator?

**Comment by DaveSkender** [comment]
@SimoneLocatelli, keep an eye on #824.  I'm currently working on the streaming quotes use case.

Currently, the library uses a more common full time-series approach, so for now you'd just add the new quote to `candles` and redo the `emaFast = candles.GetEma(21);` in your example.  This is still quite fast for normal quote histories; however, the streaming model I'm working on will be much more proficient, especially for unusually larger quote histories.

**Reply by SimoneLocatelli** [reply]
Thanks Dave, great to hear that this is currently being worked on. 
One quick question if you don't mind. I understand that indicators benefit from having at least the recommended amount of points (750 as mentioned on the library site). 
Is there a certain amount of maximum recommended points? Or the more the better? (as long that it does not impact performance too much)

Thanks for creating such a great library!

**Reply by DaveSkender** [reply]
There’s no algorithmic maximum, but you will start to observe it slow down with 10k or more quotes.  For sure, there’ll be a max that your compute and memory can handle; but I’ve not tested to that level as it’s an exceptionally rare use case.

And once you’ve met the recommended amount for an indicator, precision does not meaningfully improve further, so more quotes won’t help in that context.

Once implemented, the streaming use case will process larger datasets at a consistent faster rate that doesn’t degrade with size.

**Comment by kamrut** [comment]
> Let's suppose I create the following indicator
> 
> ```
>  List<IQuote> candles = GetAllCandles();
>  emaFast = candles.GetEma(21); 
> ```
> 
> Currently, when I add a new candle to list, I can update the indicator by recalculating it again
> 
> ```
> candles.Candle(newCandle);
> emaFast = candles.GetEma(21); 
> ```
> 
> I was wondering if this is best approach or if there is a more faster/better way to update the indicator?

The results I get from above do not match with the actual ones when I see them on chart.. I am looking to get 20 period moving average for a time frame of 15 minutes. The quotes received from the service provider match with the actual data. I am interested the moving average at date/time of the last candle closed. 

What am I missing ? Can you guide ?

List<LiveQuote> liveQuotes = GetHistoricPrices(stock, timeframe, maperiod);
IEnumerable<SmaResult> smaresults = liveQuotes.GetSma(maperiod);
SmaResult sm = smaresults.OrderByDescending(f=>f.Date).Take(1).FirstOrDefault();
return sm.Sma;

**Reply by DaveSkender** [reply]
It's hard to know what's happening without more information.  Sounds like a quotes data issue, or maybe something weird in your `GetHistoricPrices()` function.

As a side note, try `var sm= smaresults.LastOrDefault();` to simplify your query.

**Reply by kamrut** [reply]
Thanks for a prompt reply. My bad, there was an issue from the data provider, repeated results were provided for same datetime... Now used a hack to remove duplicates to conform that the averages are working correctly. Will dig more for that issue.
Many thanks for your help and above all for this wonderful library. Great Work.

**Comment by DaveSkender** [comment]
Follow or continue this conversation in our [discussion on streaming indicators](https://github.com/DaveSkender/Stock.Indicators/discussions/1018) for progress and information about pre-release package versions to support live and incremental quote sources.
